### ISSUE TITLE
(Prepare for) Handling trailing spaces in multiline strings

### ISSUE THREAD

**Issue by cgrushko** [google/google-java-format#450]
We ran into this issue with [ktfmt](https://github.com/facebookincubator/ktfmt), and solved it by replacing the last of the trailing spaces with a special tombstone to prevent GJF from removing all trailing whitespace.

https://github.com/facebookincubator/ktfmt/commit/a4fdd39237d2b59ae3c03cee9322ab04ed2b9222

I'm filing this issue so `ktfmt` can revert to using the supported method once it's available in GJF :)
(assuming multiline strings in Java are going to be a thin soon-ish)

**Comment by cushon** [comment]
I've been looking at supporting Java language features up to 14 recently, including text blocks, but hadn't thought through this yet. It makes sense that we don't want to make non-semantics-preserving changes to multiline strings. OTOH, have you encountered cases where trailing whitespace in multi-line strings is actually desired/useful? It seems like something that might deserve a style rule.

**Comment by cgrushko** [comment]
Funnily, it tripped our own tests. We have since converted to regular
strings to make it more explicit.

Do you suggest the formatter refuses to format such files?

**Comment by cushon** [comment]
> Do you suggest the formatter refuses to format such files?

No, it generally tries to accept anything that's syntactically valid and produce semantically equivalent output, I was just curious. We should probably try to preserve trailing whitespace and discourage it separately (e.g. with a linter).

**Comment by sormuras** [comment]
Closeable due to https://github.com/google/google-java-format/commit/4ddb9148c103b9c65768e519884b0c8ad5caaf39 methinks.

**Comment by cgrushko** [comment]
Cool; I didn't see a test for trailing spaces, though?

**Comment by sormuras** [comment]
> Cool; I didn't see a test for trailing spaces, though?

Dooh. I replied to the wrong issue and deleted my reply in the meanwhile.

**Comment by anthonyvdotbe** [comment]
Why isn't removing trailing white space correct, since it's removed during compilation anyway?

**Comment by cgrushko** [comment]
I don't trailing white space is removed from raw (multiline) strings during
compilation.

**Comment by anthonyvdotbe** [comment]
The algorithm described in the [JEP](http://openjdk.java.net/jeps/378#2--Incidental-white-space) has:
> 6. Remove all trailing white space [...]

For cases where trailing spaces must be kept, the line can be ended with the new `\s` escape sequence:
> \s can act as fence to prevent the stripping of trailing white space

**Comment by cgrushko** [comment]
Interesting! makes sense - otherwise removing trailing whitespace from a .java file requires parsing it.
So yeah, looks like no need to handle it in GJF.
(I don't see something similar for Kotlin, which inspired this issue; I imagine that most editors strip trailing whitespace, so in practice Kotlin programs don't have trailing whitespace)
